leader-election-impl/src/main/java/com/palantir/leader/proxy/LeadershipStateManager.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader.proxy;
+
+import com.palantir.common.concurrent.CoalescingSupplier;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.leader.LeaderElectionService.LeadershipToken;
+import com.palantir.leader.NotCurrentLeaderException;
+import com.palantir.logsafe.SafeArg;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LeadershipStateManager<T> {
+    private static final Logger log = LoggerFactory.getLogger(LeadershipStateManager.class);
+
+    private final CoalescingSupplier<LeadershipToken> leadershipTokenCoalescingSupplier;
+    private final LeadershipCoordinator leadershipCoordinator;
+    private final AtomicReference<T> delegateRef;
```

Comment:
actually - does this one need to be an AtomicReference<T> anymore? seems like it should just be a T

Reply:
I think `setDelegateToNull` will not be not be safe if we do not use AtomicRef.
    `private T setDelegateToNull() {
        T oldDelegate = delegateRef;
        delegateRef = null;
        return oldDelegate;
    }`
 If 2 threads try to clear delegate, and the delegate is refreshed in b/w, it is possible that only the newer value is returned. In that case it would be possible that we never call close() on old delegate ref.

Reply:
after reverting the change of scheduling a different thread for closing old delegate instances, I realize that `clearDelegate()` is not safe. We can have race conditions such that old delegate is never closed.